python-package/lightgbm/arrow.py

```diff
@@ -0,0 +1,57 @@
+# coding: utf-8
```

Comment:
Could you please move these definitions into `basic.py`, instead of introducing a new top-level module?

Since everything being included here is private (prefixed with `_` and excluded from `__all__`), and therefore never intended to be imported by user code, I don't think there's much benefit to separating it out into a separate module.

Reply:
Two arguments why I put it in `arrow.py`:

- `basic.py` is already rather unwieldy and I found it rather time-consuming to work with. A smaller module was much easier to handle.
- There are `pyarrow` imports in `arrow.py`: I can't easily put them in `basic.py`, hence, I currently go over `compat.py` (i.e. `basic.py` doesn't import from `.arrow` but rather from `.compat` which, in turn, imports `.arrow` in a `try-except`). I wouldn't know how to put the code in `basic.py` without adding a number of compatibility classes to `compat.py` (which I wouldn't consider particularly ergonomic).

What about naming the file `_arrow.py` to make it clear that it is a private module?

Reply:
> ... adding a number of compatibility classes to `compat.py` (which I wouldn't consider particularly ergonomic)

This is exactly what I'd like you to do, please.

Based on the current state of this PR, you're only using 1 class (`pyarrow.Table`) and 5 functions (`pyarrow.types.is_floating()`, `pyarrow.types.is_integer()`, `pyarrow.cffi.addressof()`, `pyarrow.cffi.cast()`, `pyarrow.cffi.new()`). That's not so many imports that I think it is worth deviating from the patterns `lightgbm` uses for interacting with external dependencies. There is also "ergonomic" value in similar operations being handled in similar ways throughout the codebase.

> I wouldn't know how to put the code in `basic.py` ...

Put this in `compat.py`:

```python
"""pyarrow"""
try:
    from pyarrow import Table as pa_Table
    from pyarrow.cffi import (
        addressof AS arrow_ffi_addressof,
        cast AS arrow_ffi_cast,
        new AS arrow_ffi_new,
    )
    from pyarrow.types import (
        is_floating as arrow_is_floating,
        is_integer as arrow_is_integer
    )
    PYARROW_INSTALLED = True
except ImportError:
    PYARROW_INSTALLED = False

    class pa_Table:  # type: ignore
        """Dummy class for pa.Table."""

        def __init__(self, *args, **kwargs):
            pass

    arrow_is_integer = None
    arrow_is_floating = None
    arrow_ffi_addressof = None
    arrow_ffi_cast = None
    arrow_ffi_new = None
```

And then in `basic.py`


```python
from .compat import arrow_is_integer, arrow_is_floating, arrow_ffi_addressof, arrow_ffi_cast, arrow_ffi_new

if TYPE_CHECKING:
    from pyarrow.cffi import CData as pyarrow_CData

# ...
class _ArrowCArray:
    """Simple wrapper around the C representation of an Arrow type."""

    n_chunks: int
    chunks: "pyarrow_CData"
    schema: "pyarrow_CData"

    def __init__(self, n_chunks: int, chunks: "pyarrow_CData", schema: "pyarrow_CData"):
        self.n_chunks = n_chunks
        self.chunks = chunks
        self.schema = schema

    @property
    def chunks_ptr(self) -> int:
        """Returns the address of the pointer to the list of chunks making up the array."""
        return int(ffi.cast("uintptr_t", ffi.addressof(self.chunks[0])))

    @property
    def schema_ptr(self) -> int:
        """Returns the address of the pointer to the schema of the array."""
        return int(ffi.cast("uintptr_t", self.schema))
```

If you're not familiar with that type-hints-in-a-string approach in Python, see https://mypy.readthedocs.io/en/stable/runtime_troubles.html.

Reply:
Thanks for the elaborate explanation! Done ✅